system-addon/test/functional/mochitest/browser_as_load_location.js

```diff
@@ -1,34 +1,48 @@
 "use strict";
 
-let Cu = Components.utils;
-Cu.import("resource://gre/modules/Services.jsm");
-
-/**
- * Tests that opening a new tab opens a page with the expected activity stream
- * content.
- *
- * XXX /browser/components/newtab/tests/browser/browser_newtab_overrides in
- * mozilla-central is where this test was adapted from.  Once we get decide on
- * and implement how we're going to set the URL in mozilla-central, we may well
- * want to (separately from this test), clone/adapt that entire file for our
- * new setup.
- */
-add_task(async function checkActivityStreamLoads() {
-  const asURL = "resource://activity-stream/data/content/activity-stream.html";
-
+// Helper to test that a newtab page loads
+async function checkNewtabLoads(selector, message) {
   // simulate a newtab open as a user would
   BrowserOpenTab();
 
   // wait until the browser loads
   let browser = gBrowser.selectedBrowser;
-  await BrowserTestUtils.browserLoaded(browser);
+  await waitForPreloaded(browser);
 
   // check what the content task thinks has been loaded.
-  await ContentTask.spawn(browser, {url: asURL}, args => {
-    Assert.ok(content.document.querySelector("body.activity-stream"),
-      'Got <body class="activity-stream" Element');
-  });
+  let found = await ContentTask.spawn(browser, selector, arg =>
+    content.document.querySelector(arg) !== null);
+  ok(found, message);
 
   // avoid leakage
   await BrowserTestUtils.removeTab(gBrowser.selectedTab);
+}
+
+// Test with activity stream on
+async function checkActivityStreamLoads() {
+  await checkNewtabLoads("body.activity-stream", "Got <body class='activity-stream'> Element");
+}
+
+// Run a first time not from a preloaded browser
+add_task(async function checkActivityStreamNotPreloadedLoad() {
+  gBrowser.removePreloadedBrowser();
+  await checkActivityStreamLoads();
+});
+
+// Run a second time from a preloaded browser
+add_task(checkActivityStreamLoads);
+
+// Test with activity stream off
+async function checkNotActivityStreamLoads() {
+  await SpecialPowers.pushPrefEnv({set: [["browser.newtabpage.activity-stream.enabled", false]]});
```

Comment:
Does this need to be popped manually in cleanup somewhere, since it's not in a scopedPushPrefs?